### ISSUE TITLE
Proof of concept - plantuml support

### ISSUE THREAD

**Issue by alexfargus** [brechtm/rinohtype#272]
I need plantuml support for my documents so I picked up the trail from #155.

I have had some some success using the png output from plantuml. (See [plantuml.pdf](https://github.com/brechtm/rinohtype/files/6744003/plantuml.pdf)) The problem encountered by @vtexier is due to an off py one error in `purepng.py` that causes an assertion to fail rather than skipping compressed metadata as intended. 

@brechtm  I'll head upstream to see if this issue has been fixed elsewhere. However since `purepng.py` is embedded into rinohtype will you accept a PR to fix the issue here?

The changes from my proof of concept branch can be found [here](https://github.com/brechtm/rinohtype/commit/024b57db29387bb2ca08d914df08ea5c37415856)

With regards to the Tansformer used to support the plantuml node, I don't think that it fits in rinohtype since that would introduce a dependency on `sphinxcontrib-plantuml` so I will find another home for that.

**Comment by brechtm** [comment]
> I have had some some success using the png output from plantuml. (See [plantuml.pdf](https://github.com/brechtm/rinohtype/files/6744003/plantuml.pdf)) The problem encountered by @vtexier is due to an off py one error in `purepng.py` that causes an assertion to fail rather than skipping compressed metadata as intended.
> 
> @brechtm I'll head upstream to see if this issue has been fixed elsewhere. However since `purepng.py` is embedded into rinohtype will you accept a PR to fix the issue here?

Thanks for debugging! Yes, please submit this fix upstream. I see they have a test suite, so that should catch regressions due to the change. There doesn't seem to be much activity over at purepng's GitHub page though, so if your fix passes the test suite, I'm happy to include it in rinohtype.

> The changes from my proof of concept branch can be found [here](https://github.com/brechtm/rinohtype/commit/024b57db29387bb2ca08d914df08ea5c37415856)
>
> With regards to the Tansformer used to support the plantuml node, I don't think that it fits in rinohtype since that would introduce a dependency on sphinxcontrib-plantuml so I will find another home for that.

Yup. Except for the purepng fix, all code should become part of the plantuml codebase. You'll have to making sure it still works if _rinoh_ can't be imported, for example by importing rinoh within by a try/except.

I don't think you need a transformer though. Did you see the example code from https://github.com/brechtm/rinohtype/issues/118#issuecomment-613582610? That is rinohtype's equivalent of the visit/depart functions of other builders.

PS. I haven't yet found the time to review #271. Will do that ASAP.

**Comment by alexfargus** [comment]
> I don't think you need a transformer though. Did you see the example code from [#118 (comment)](https://github.com/brechtm/rinohtype/issues/118#issuecomment-613582610)? That is rinohtype's equivalent of the visit/depart functions of other builders.

Yes, I saw that. However in my opinion the advantage of the transformer approach is that it *should* work for any builder, and not need to import anything specific to rinohtype (my implementation only uses imports from sphinx and plantuml).

**Comment by brechtm** [comment]
> However in my opinion the advantage of the transformer approach is that it _should_ work for any builder, and not need to import anything specific to rinohtype (my implementation only uses imports from sphinx and plantuml).

That's a good point.

For completeness, in many cases you can simply return a tree of standard docutils nodes from the _SphinxDirective_'s subclass's `run()` method (no need for visit/depart methods). You could even do things specific to the builder in that method. See https://github.com/brechtm/rinohtype/issues/201#issuecomment-736334190 and https://github.com/brechtm/rinohtype/issues/201#issuecomment-737534291 for more info, and the [`DrawIOImage` and `DrawIOFigure` classes](https://github.com/Modelmat/sphinxcontrib-drawio/blob/76b9a506a8c65950a305bc50577576ec7f6438d9/sphinxcontrib/drawio/__init__.py#L70) for an example.

**Comment by brechtm** [comment]
I noticed that PyPNG, where purepng was forked from, is again being updated. It might be good to submit the fix to PyPNG (if it still has the same bug) and see whether we should replace purepng with it.

**Comment by alexfargus** [comment]
PyPng does not have this bug. This is due to the fact that it does not try to parse the infrmation that is mishandled in purepng.

See https://github.com/drj11/pypng/blob/45e8e314ca0bf1cd0c390ac95d353c9b31b17723/man/chunk.rst#itxt

**Comment by basejumpa** [comment]
👍 for that feature. I also need it!

I'd love if this enhancement would support both png and svg formats.
